### PR TITLE
Introduce GradleDistributionsPlugin for gradle/gradle build to model different distributions for integration testing

### PR DESCRIPTION
**Pull request by jjohannes**
We can now have different distributions, containing different sets of jars/functionality as proposed [here](https://github.com/gradle/gradle-private/issues/1121#issue-300327577). For this, one subproject per distribution type is created. Initially this PR introduces
  - _core_ (absolute minimum needed to run Gradle)
  - _basics_ (Basic functionalities on top of core)
  - _jvm_ (Everything needed for JVM ecosystem support)
  - _native_ (everything for Native support)
  - _publishing_ (everything to publish to Maven/Ivy repositories)
  - _full_ (everything - what we publish!)

This is just an initial structure which we can now easily change. The main point of this PR is to enable a setup like this at all. New/different distributions can now simply be added by adding a new subproject that applies `gradlebuild.distribution.packaging` and defines the content of the distribution in terms of dependencies.
Projects needing the distribution for testing, then depend on that project - e.g. `integTestDistributionRuntimeOnly(project(":distributionsJvm"))`

While introducing this, a lot of changes/improvements are done. (See also commits and documentation in code.) Some major things:

- The embedded test runner now no longer supports running against a full distribution as fallback. This has the advantage that we do not need to assemble the distribution in embedded running, where the huge majority of tests does not need it. To allow more tests still to run embedded, the embedded support with TestKit/KotlinDSL involvement has been improved. Some tests won't run with the embedded runner at all now, but compared to the overall amount of tests, these are just a few. And the "forking" tests always run as part of the PR pipeline.
- Cacheability of tests is highly improved. Because there are so much changes to test inputs in this PR, I checked this carefully and discovered that we broke some things recently when adding inputs for distributed testing. These regressions are addressed. Other things were improved along the line (local Tooling API repo no longer uses timestamp version, local bin distribution without timestamp for wrapper tests, sample tests no longer rely on absolute path and timestamp version, ...) 
- The `docs.jar` is no longer part of the distribution. It almost had no content and what it had was runtime metadata and not documentation. Instead, a distribution project now encapsulates all the runtime metadata creation and produces a `runtime-api-info.jar` that is specific to the distribution.
- In the past we had flakiness issues causes by different distributions sharing the same user home/daemon registry. With the fixed set of distributions defined, we now use different user home/daemon registry for each distribution.
- Cross-project configurations are removed. There was some code central to distribution building and runtime metadata generation that reached into all subprojects. This is now replaced by using proper variant aware dependency management to share artifacts between distribution subprojects and the other subprojects.

### CODE REVIEW

**Comment by jjohannes** [comment]
@bot-gradle test his

**Comment by jjohannes** [comment]
@bot-gradle test this

**Comment by jjohannes** [comment]
@bot-gradle test RFR

**Comment by jjohannes** [comment]
@bot-gradle test RFR

**Comment by jjohannes** [comment]
@bot-gradle test this

File: subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy [hunk 1]
```diff
@@ -1,61 +0,0 @@
...
-
-package org.gradle.internal.nativeintegration.jansi
-
-import org.gradle.util.Requires
-import spock.lang.Specification
-
-import static org.gradle.internal.nativeintegration.jansi.JansiLibraryFactory.*
-import static org.gradle.util.TestPrecondition.*
-
-class JansiLibraryFactoryIntegrationTest extends Specification {
```
Comment (wolfs) [inline subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy:25]: Why did you remove this test without replacement?

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt [hunk 2]
```diff
@@ -75,7 +75,7 @@ open class BuildCommitDistribution : DefaultTask() {
             "--init-script",
             File(checkoutDir, "gradle/init-scripts/build-scan.init.gradle.kts").absolutePath,
             "clean",
-            ":distributions:install",
+            "install",
```
Comment (wolfs) [inline buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt:78]: Has it been fixed that `install` also installs stuff into the local `.m2` directory?

Comment (jjohannes) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt:78]: Sorry I am not sure what you are talking about. Can you point me to some description of what is broken/what used to work?

File: subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy (hunk 1 above)
Comment (jjohannes) [reply subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy:25]: This was not an integration tests in the sense that it actually ran Gradle. So I wanted to move this into `test` as I thought it would be better placed there. To avoid this to be executed several times on CI. But then I saw that the followin test is (almost) the same so I removed this one as "duplicate".

https://github.com/gradle/gradle/blob/master/subprojects/native/src/test/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryTest.groovy

But rechecking now, I see that there is a slight difference in the tests. So I should probably add this back. Thanks for asking.

Would you agree that it should go into `test` then (at least for now until we might add another set for "normal integration tests")?

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt (hunk 2 above)
Comment (wolfs) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt:78]: Nothing seems to be broken. I remember from some time ago that running `install` without specifying a project caused running the `install` task from the `maven` plugin. I suppose now that we don't apply the `maven` plugin any more, there is also no `install` task to run for every project, right?

I think it still would prefer to qualify which `install` task to run here, so to use `:distributions:install`. Why did you change the invocation to run just `install`?

Comment (jjohannes) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt:78]: Got it. That problem does not exist anymore since we moved to the "maven-publush" plugin. (I wasn't aware myself that we had it before, but it makes sense what you say.)

So locally you can now safely run `install`, which is nice as the full qualified name now became longer with recent changes.

I changed it here because it now became `distributionFull:install`. And because this builds the master version of Gradle, the change can only be made after the merge, if you want CI to pass on the branch.

But now that this is merged, I can change it to the fully qualified name again for clarity on master directly.

File: subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy:25]: Running this as a normal test is perfectly fine.

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt (hunk 2 above)
Comment (wolfs) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt:78]: Got it. Thank you for the explanation.

File: subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy (hunk 1 above)
Comment (jjohannes) [reply subprojects/native/src/integTest/groovy/org/gradle/internal/nativeintegration/jansi/JansiLibraryFactoryIntegrationTest.groovy:25]: https://github.com/gradle/gradle/commit/a2cc431d4165eb53338a6e6cbe5db43aead2e3d1

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt (hunk 2 above)
Comment (jjohannes) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/BuildCommitDistribution.kt:78]: https://github.com/gradle/gradle/commit/3ef0726b2a184e8fdb493f4bc819cc395cf86451
